x-pack/platform/plugins/shared/cases/server/services/incremental_id/index.ts

```diff
@@ -0,0 +1,391 @@
+        }
+
+        const caseSo = casesWithoutIncrementalId[index];
+        const namespaceOfCase = caseSo.namespaces?.[0];
+        if (!namespaceOfCase) {
+          this.logger.error(`Case ${caseSo.id} has no namespace assigned. Skipping it.`);
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+
+        // Get the incremental id SO from the cache or fetch it
+        let incIdSo = incIdSoCache.get(namespaceOfCase);
+        if (!incIdSo) {
+          this.logger.debug(
+            `Don't have incrementer in cache, fetching it: namespace ${namespaceOfCase}`
+          );
+          incIdSo = await this.getOrCreateCaseIdIncrementerSo(namespaceOfCase);
+          this.logger.debug(
+            `Fetched incrementer SO for ${namespaceOfCase}: ${JSON.stringify(incIdSo)}`
+          );
+          incIdSoCache.set(namespaceOfCase, incIdSo);
+        }
+
+        // Increase the inc id
+        const newId = incIdSo.attributes.last_id + 1;
+        // Apply the new ID to the case
+        await this.applyIncrementalIdToCaseSo(caseSo, newId, namespaceOfCase);
+        // Apply the new ID to the local incrementer SO, it will persist later
+        incIdSo.attributes.last_id = newId;
+        hasAppliedAnId = true;
+        countProcessedCases++;
+      } catch (error) {
+        this.logger.error(`ID incrementing paused due to error: ${error}`);
+        break;
+      }
+    }
+
+    // If changes have been made, apply the changes to the counters
+    // These are done in sequence, since we cannot guarantee that `incIdSoCache` is small.
+    // It might have hundreds/thousands of SO objects cached that need updating.
+    if (hasAppliedAnId && !this.isStopped) {
+      for (const [namespace, incIdSo] of incIdSoCache) {
+        await this.incrementCounterSO(incIdSo, incIdSo.attributes.last_id, namespace);
+      }
+    }
+
+    return countProcessedCases;
+  }
+
+  getCaseIdIncrementerSo(namespace: string) {
+    return this.internalSavedObjectsClient.find<CaseIdIncrementerPersistedAttributes>({
+      type: CASE_ID_INCREMENTER_SAVED_OBJECT,
+      namespaces: [namespace],
+    });
+  }
+
+  /**
+   * Gets or creates the case id incrementer SO for the given namespace
+   * @param namespace The namespace of the case id incrementor so
+   */
+  async getOrCreateCaseIdIncrementerSo(
+    namespace: string
+  ): Promise<SavedObject<CaseIdIncrementerPersistedAttributes>> {
+    try {
+      const [latestAppliedId, incrementerResponse] = await Promise.all([
+        // Get the latest applied id by looking at the case saved objects
+        await this.getLastAppliedIdForSpace(namespace),
+        // Get the case id incrementer saved object
+        await this.getCaseIdIncrementerSo(namespace),
+      ]);
+      this.logger.debug(`Latest applied ID to a case for ${namespace}: ${latestAppliedId}`);
+
+      const actualLatestId = latestAppliedId || 0;
+
+      const incrementerSO = incrementerResponse?.saved_objects[0];
```

Comment:
Though unlikely, what happens if the incrementer SO doesn't exist for whatever reason?

Reply:
Then the `else` case kicks in:

```js
// At this point we assume that no incrementer SO exists
this.logger.debug(`No incrementer found for ${namespace}. Creating a new one.`);
return this.createCaseIdIncrementerSo(namespace);
```